### DISCUSSION TITLE
Define a geometry as an ordered list of coordinates in a column?

### DISCUSSION THREAD

**Discussion by fox-daniel** [geoarrow/geoarrow#54] [Q&A] [answered]
Has GeoArrow considered supporting a geometry format that is more columnar oriented? I wonder if this could help speed up some common geospatial algorithms since there seems to be a lot of support for columnar oriented computing. The kind of geometry format I have in mind is for the file format to have columns for x, y, z coordinates and attributes, no geometry column. A geometry would then be defined by an attribute column that assigned a GeoID to all the rows. The GeoID could have a prefix that indicates which geometry the coordinate is an element of, and a suffix that indicates the order of the points and whether they are part of an inner or outer ring.

This could also possibly help with file size (though only be roughly a factor of 2). A lot of geometry files have points that are shared by multiple geometries. For example if you are looking at country boundaries, most points appear in a least two geometries and some points appear in three or more. Maybe compression could be used that only stores the relatively large floats of the coordinates for each point once but links them to the multiple sets of attributes in which they are used.

I imagine this has been considered and dismissed or is part of another project. I'd love to hear why this is not a good approach or which format is already using it. Thanks for any thoughts!

**Comment by kylebarron** [comment, marked as answer]
> more columnar oriented

I think maybe this is dependent on your definition of column-oriented. Right now all coordinates for a column are stored in a contiguous array, which seems maximally column-oriented, even if there are a few different approaches to columnar storage.

> The GeoID could have a prefix that indicates which geometry the coordinate is an element of, and a suffix that indicates the order of the points and whether they are part of an inner or outer ring. This could also possibly help with file size (though only be roughly a factor of 2).

You have to weigh the additional space used for the lookup array against the space saved by deduplicating the coordinates, _and_ the CPU time it takes to find the duplicated coordinates. This seems like an optimization for a very specific type of data (i.e. polygonal coverages) like TopoJSON that might be less efficient for general data without duplicated coordinates.

> This could also possibly help with file size

GeoArrow is really not focused on file size. GeoArrow is focused on efficient memory access with good cache locality, zero-copy FFI, and being compatible with the entire Arrow ecosystem.

Your layout is probably very similar to the RLE encoding that Parquet does automatically on the offsets array when serializing Arrow to Parquet.

**Reply by fox-daniel** [reply]
Thanks for clarifying. I reread the documentation after your comments and see that the Coordinate (separated) memory layout already does what I was interested in. Thanks for pointing out the RLE encoding from Parquet. That covers my other question!
